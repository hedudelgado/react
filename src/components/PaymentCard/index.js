export { default } from './PaymentCard';

export { default as PaymentCardContainer } from './PaymentCardContainer';
