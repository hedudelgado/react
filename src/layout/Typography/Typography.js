import React from 'react';
import {
  string, node, oneOf,
} from 'prop-types';
import classNames from 'classnames';

import BaseStyle from '../BaseStyle';

const weights = {
  bold: 'bold',
};

const availableSizes = {
  xxxl: 'xxxl',
  xxl: 'xxl',
  xl: 'xl',
  l: 'l',
  m: 'm',
  base: 'base',
  s: 's',
  xs: 'xs',
};

const headerComponentTypes = {
  hero: 'h1',
  h1: 'h1',
  h2: 'h2',
  h3: 'h3',
  h4: 'h4',
  h5: 'h5',
  h6: 'h6',
  subtext: 'h5',
};

const textComponentTypes = {
  p: 'p',
  span: 'span',
};

const componentTypes = {
  ...headerComponentTypes,
  ...textComponentTypes,
};

const Typography = ({
  children,
  component,
  size = availableSizes.base,
  className: classNameProp,
  weight,
  ...otherProps
}) => {
  const className = classNames(
    {
      'wb-heading': headerComponentTypes[component],
      [`wb-heading--${component}`]: headerComponentTypes[component],
      [`font-size--${size}`]: textComponentTypes[component],
      'wb-strong': weights[weight] && weight === weights.bold,
    },
    classNameProp,
  );
  return (
    <BaseStyle
      component={componentTypes[component]}
      className={className}
      {...otherProps}
    >
      {
        children
      }
    </BaseStyle>
  );
};

Typography.propTypes = {
  /**
   * Either string or react node
   */
  children: node,
  /**
   * All possible components
   */
  component: oneOf(Object.keys(componentTypes)),
  /* eslint-disable react/require-default-props */
  size: oneOf(Object.keys(availableSizes)),
  className: string,
  weight: string,
};

Typography.defaultProps = {
  component: componentTypes.p,
};

export default Typography;
