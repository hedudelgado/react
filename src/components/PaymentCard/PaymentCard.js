import React, { Fragment } from 'react';
import { func, object } from 'prop-types';
import {
  DEFAULT_VIEW,
  PAYMENT_CARD_VIEW,
  PAYMENT_CARD_ERROR,
} from '../App/appActions';
import { getJsonItem } from '../../utils/session-storage';
import Button from '../../layout/Button';
import Typography from '../../layout/Typography';
import BaseStyle from '../../layout/BaseStyle';
import DeleteModal from '../../layout/DeleteModal';
import Notification from '../../layout/Notification';
import { paymentCards } from '../../config.json';

const PaymentCard = ({
  app: { notificationMessage },
  setEditView,
  setNotificationMessage,
  dictionary,
  openModal,
  closeModal,
  profile,
  updateProfileStore,
  updateProfile,
}) => {
  const deleteCard = () => {
    const payload = {
      ...profile,
      paymentPreference: {
        ...profile.paymentPreference,
        paymentCard: {},
      },
    };

    updateProfile(payload)
      .then(() => {
        updateProfileStore(payload);
        closeModal();
        setNotificationMessage(PAYMENT_CARD_VIEW);
      }).catch(() => {
        setNotificationMessage(PAYMENT_CARD_ERROR);
      });
  };

  const showEditMode = () => {
    setEditView(PAYMENT_CARD_VIEW);
    setNotificationMessage(DEFAULT_VIEW);
  };
  const {
    paymentPreference: {
      paymentCard: {
        cardType = '',
        cardNumber = '',
        expiryDate = '',
        cardHolderName = '',
      } = {},
    } = {},
  } = profile;
  const splitExpiryDate = expiryDate.split('/');
  const isCardExpired = new Date(`20${splitExpiryDate[1]}`, splitExpiryDate[0]) < new Date();
  const formattedCardNumber = `•••• •••• •••• ${cardNumber.replace(/\*/g, '')}`;

  return (
    <section className="component-container">
      {notificationMessage === PAYMENT_CARD_ERROR && (
        <Notification data-test="error-notification" priority={2} variant="error" ariaLive="polite" show>
          {dictionary['generic.error']}
        </Notification>
      )}
      {notificationMessage === PAYMENT_CARD_VIEW && (
        <Notification data-test="success-message" priority={2} variant="success" ariaLive="polite" show>
          {dictionary['update.success.message']}
        </Notification>
      )}
      <Typography component="h4" className="pi-color-greyscale-darkest">
        {dictionary['payment.title']}
      </Typography>
      {isCardExpired && (
        <div data-test="error" className="wb-notification wb-notification__unhousedalert">
          <i className="wb-notification-icon wb-icon_disc-info-new" />
          <div className="wb-notification-text">
            {dictionary['payment.expired']}
          </div>
        </div>
      )}
      {!!cardType && (
        <BaseStyle className="basic-spacer" component="div" mb="3">
          <div className="card-layout p2 mt1 border rounded wb-push--bottom">
            <div className="right-align bold pb1">{paymentCards[cardType].name}</div>
            <div className="right-align">{formattedCardNumber}</div>
            <div className="clearfix col-12 card-details relative">
              <div className="col col-10">
                <div className="pi-color-greyscale-dark">
                  {dictionary['payment.expirydate']}
                </div>
                <div className="pb1">{expiryDate}</div>
                <div>{cardHolderName}</div>
              </div>
              <div className="col col-2 right-align">
                <img
                  src={getJsonItem('environment').domain + paymentCards[cardType].img}
                  className="block right absolute bottom-0 right-0"
                  alt={paymentCards[cardType].name}
                />
              </div>
            </div>
          </div>
          <Button
            data-test="deleteCard"
            className="wb-text-btn--underline block wb-push--bottom"
            variant="text"
            color="red"
            onClick={() => openModal(
              <DeleteModal
                title={dictionary['payment.delete.title']}
                subtext={dictionary['payment.delete.text']}
                deleteFunc={deleteCard}
                buttonLabel={dictionary['payment.delete.button']}
                handleClose={closeModal}
              />,
            )}
          >
            {dictionary['payment.button.delete']}
          </Button>
          <Button id="edit-card" color="quaternary" onClick={showEditMode}>
            {dictionary['payment.button.addnew']}
          </Button>
        </BaseStyle>
      )}
      {!cardType && (
        <Fragment>
          <Typography mb="3" data-test="no-guests-message">
            {dictionary['payment.nocard']}
          </Typography>
          <Button id="add-card" color="secondary" fullWidth onClick={showEditMode}>
            {dictionary['payment.button.add']}
          </Button>
        </Fragment>
      )}
    </section>
  );
};

PaymentCard.propTypes = {
  app: object.isRequired,
  profile: object.isRequired,
  dictionary: object.isRequired,
  setEditView: func.isRequired,
  setNotificationMessage: func.isRequired,
  openModal: func.isRequired,
  closeModal: func.isRequired,
  updateProfileStore: func.isRequired,
  updateProfile: func.isRequired,
};

export default PaymentCard;
