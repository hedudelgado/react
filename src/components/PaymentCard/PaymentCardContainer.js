import { connect } from 'react-redux';
import PaymentCard from './PaymentCard';
import { openModal, closeModal } from '../App/modalActions';
import {
  updateUserProfile as updateProfile,
} from '../App/appActions';

const mapDispatchToProps = dispatch => ({
  openModal: content => dispatch(openModal(content)),
  closeModal: () => dispatch(closeModal()),
  updateProfile: params => dispatch(updateProfile(params)),
});

export default connect(null, mapDispatchToProps)(PaymentCard);
