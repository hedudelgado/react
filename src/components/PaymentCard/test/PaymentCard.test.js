import React from 'react';
import { shallow } from 'enzyme';
import {
  DEFAULT_VIEW,
  PAYMENT_CARD_ERROR,
  PAYMENT_CARD_VIEW,
} from '../../App/appActions';
import PaymentCard from '../PaymentCard';
import Typography from '../../../layout/Typography';

jest.mock('../../../utils/session-storage', () => ({
  __esModule: true,
  getJsonItem: (key) => {
    switch (key) {
      case 'domain':
        return 'https://www.premierinn.com';

      default:
        return '';
    }
  },
}));

describe('PaymentCard Component', () => {
  const updateProfile = jest.fn(() => Promise.resolve({}));
  let wrapper;

  const props = {
    updateProfile,
    profile: {
      paymentPreference: {
        electronicInvoiceRequired: true,
        paymentCard: {
          cardType: 'VI',
          cardNumber: '************1111',
          expiryDate: '07/23',
          cardHolderName: 'dwa',
        },
      },
    },
    dictionary: {
      'update.success.message': 'Your details have been updated successfully!',
      'payment.nocard': 'Save your card details for quicker, easier booking next time.',
      'payment.button.add': 'Add payment card',
      'payment.button.delete': 'Delete card',
      'payment.button.addnew': 'Add new card',
      'payment.expirydate': 'Expiry date',
      'payment.expired': 'This card has expired.',
      'payment.title': 'Payment',
      'payment.delete.title': 'Delete payment method',
      'payment.delete.text': 'Are you sure you want to delete this payment method?',
      'payment.delete.button': 'Delete payment card',
    },
    app: {},
    setEditView: jest.fn(),
    setNotificationMessage: jest.fn(),
    openModal: jest.fn(),
    closeModal: jest.fn(),
    updateProfileStore: jest.fn(),
  };

  beforeEach(() => {
    jest.clearAllMocks();
    wrapper = shallow(<PaymentCard {...props} />);
  });

  describe('Renders PaymentCard Component Elements', () => {
    it('should render the component', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should render the edit card button', () => {
      expect(wrapper.find('#edit-card')).toHaveLength(1);
    });

    it('should show success message when payment card updated successfully', () => {
      const copyProps = {
        ...props,
        app: { editViewName: DEFAULT_VIEW, notificationMessage: PAYMENT_CARD_VIEW },
      };
      wrapper = shallow(<PaymentCard {...copyProps} />);
      expect(wrapper).toContainExactlyOneMatchingElement('[data-test="success-message"]');
    });

    it('should show error notification', () => {
      const copyProps = {
        ...props,
        app: { editViewName: DEFAULT_VIEW, notificationMessage: PAYMENT_CARD_ERROR },
      };
      wrapper = shallow(<PaymentCard {...copyProps} />);
      expect(wrapper).toContainExactlyOneMatchingElement('[data-test="error-notification"]');
    });

    it('should show card expired message when card expiry date is in the past', () => {
      const copyProps = {
        ...props,
        profile: {
          paymentPreference: {
            paymentCard: {
              cardType: 'VI',
              cardNumber: '************1111',
              expiryDate: '07/00',
              cardHolderName: 'dwa',
            },
          },
        },
      };
      wrapper = shallow(<PaymentCard {...copyProps} />);
      expect(wrapper.contains(<div className="wb-notification-text">This card has expired.</div>)).toBeTruthy();
    });

    it('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should show no card message', () => {
      const copyProps = {
        ...props,
        profile: {
          paymentPreference: {},
        },
      };
      wrapper = shallow(<PaymentCard {...copyProps} />);
      expect(wrapper.contains(
        <Typography mb="3" data-test="no-guests-message">
          Save your card details for quicker, easier booking next time.
        </Typography>,
      )).toBeTruthy();
    });
  });

  describe('Triggers the correct actions', () => {
    it('should trigger setEditView when clicking the change password button', () => {
      wrapper.find('#edit-card').simulate('click');
      expect(props.setEditView).toHaveBeenCalled();
      expect(props.setNotificationMessage).toHaveBeenCalled();
    });

    it('should open the delete modal on click', () => {
      wrapper.find('[data-test="deleteCard"]').simulate('click');
      expect(props.openModal).toHaveBeenCalledTimes(1);
    });

    it('should set ERROR notification if api fails', async () => {
      updateProfile.mockImplementationOnce(jest.fn(() => Promise.reject({})));
      props.openModal = jest.fn(({ props: { deleteFunc } }) => deleteFunc());
      wrapper = shallow(<PaymentCard {...props} />);
      wrapper.find('[data-test="deleteCard"]').simulate('click');
      expect(props.openModal).toHaveBeenCalledTimes(1);

      await Promise.all([updateProfile]).then(() => {
        expect(props.setNotificationMessage)
          .toHaveBeenCalledTimes(1)
          .toHaveBeenCalledWith(PAYMENT_CARD_ERROR);
        expect(props.updateProfileStore).not.toHaveBeenCalled();
      });
    });

    it('should delete payment card', async () => {
      props.openModal = jest.fn(({ props: { deleteFunc } }) => deleteFunc());
      wrapper = shallow(<PaymentCard {...props} />);
      wrapper.find('[data-test="deleteCard"]').simulate('click');
      expect(props.openModal).toHaveBeenCalledTimes(1);

      await Promise.all([updateProfile]).then(() => {
        expect(updateProfile).toHaveBeenCalledTimes(1);
        expect(updateProfile).toHaveBeenCalledWith({
          paymentPreference: {
            electronicInvoiceRequired: true,
            paymentCard: {},
          },
        });
      });

      expect(props.closeModal).toHaveBeenCalledTimes(1);
      expect(props.setNotificationMessage).toHaveBeenCalledTimes(1);
      expect(props.updateProfileStore).toHaveBeenCalledTimes(1);
      expect(props.updateProfileStore).toHaveBeenCalledWith({
        paymentPreference: {
          electronicInvoiceRequired: true,
          paymentCard: {},
        },
      });
    });
  });

  describe('PaymentCard form default values', () => {
    it('should default to add payment card view ', () => {
      const paymentCardProps = {
        ...props,
      };

      delete paymentCardProps.profile.paymentPreference;

      const paymentCardWrapper = shallow(<PaymentCard {...paymentCardProps} />);
      expect(paymentCardWrapper).toMatchSnapshot();
    });
  });
});
