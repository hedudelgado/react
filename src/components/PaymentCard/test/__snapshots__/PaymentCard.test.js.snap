// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaymentCard Component PaymentCard form default values should default to add payment card view  1`] = `
<section
  className="component-container"
>
  <Typography
    className="pi-color-greyscale-darkest"
    component="h4"
  >
    Payment
  </Typography>
  <Typography
    component="p"
    data-test="no-guests-message"
    mb="3"
  >
    Save your card details for quicker, easier booking next time.
  </Typography>
  <Button
    color="secondary"
    fullWidth={true}
    href={null}
    id="add-card"
    onClick={[Function]}
    submit={false}
    variant="button"
  >
    Add payment card
  </Button>
</section>
`;

exports[`PaymentCard Component Renders PaymentCard Component Elements should match snapshot 1`] = `
<section
  className="component-container"
>
  <Typography
    className="pi-color-greyscale-darkest"
    component="h4"
  >
    Payment
  </Typography>
  <BaseStyle
    className="basic-spacer"
    component="div"
    mb="3"
  >
    <div
      className="card-layout p2 mt1 border rounded wb-push--bottom"
    >
      <div
        className="right-align bold pb1"
      >
        Visa Credit
      </div>
      <div
        className="right-align"
      >
        •••• •••• •••• 1111
      </div>
      <div
        className="clearfix col-12 card-details relative"
      >
        <div
          className="col col-10"
        >
          <div
            className="pi-color-greyscale-dark"
          >
            Expiry date
          </div>
          <div
            className="pb1"
          >
            07/23
          </div>
          <div>
            dwa
          </div>
        </div>
        <div
          className="col col-2 right-align"
        >
          <img
            alt="Visa Credit"
            className="block right absolute bottom-0 right-0"
            src="undefined/content/dam/global/booking/VC.jpg"
          />
        </div>
      </div>
    </div>
    <Button
      className="wb-text-btn--underline block wb-push--bottom"
      color="red"
      data-test="deleteCard"
      fullWidth={false}
      href={null}
      onClick={[Function]}
      submit={false}
      variant="text"
    >
      Delete card
    </Button>
    <Button
      color="quaternary"
      fullWidth={false}
      href={null}
      id="edit-card"
      onClick={[Function]}
      submit={false}
      variant="button"
    >
      Add new card
    </Button>
  </BaseStyle>
</section>
`;

exports[`PaymentCard Component Renders PaymentCard Component Elements should render the component 1`] = `
<section
  className="component-container"
>
  <Typography
    className="pi-color-greyscale-darkest"
    component="h4"
  >
    Payment
  </Typography>
  <BaseStyle
    className="basic-spacer"
    component="div"
    mb="3"
  >
    <div
      className="card-layout p2 mt1 border rounded wb-push--bottom"
    >
      <div
        className="right-align bold pb1"
      >
        Visa Credit
      </div>
      <div
        className="right-align"
      >
        •••• •••• •••• 1111
      </div>
      <div
        className="clearfix col-12 card-details relative"
      >
        <div
          className="col col-10"
        >
          <div
            className="pi-color-greyscale-dark"
          >
            Expiry date
          </div>
          <div
            className="pb1"
          >
            07/23
          </div>
          <div>
            dwa
          </div>
        </div>
        <div
          className="col col-2 right-align"
        >
          <img
            alt="Visa Credit"
            className="block right absolute bottom-0 right-0"
            src="undefined/content/dam/global/booking/VC.jpg"
          />
        </div>
      </div>
    </div>
    <Button
      className="wb-text-btn--underline block wb-push--bottom"
      color="red"
      data-test="deleteCard"
      fullWidth={false}
      href={null}
      onClick={[Function]}
      submit={false}
      variant="text"
    >
      Delete card
    </Button>
    <Button
      color="quaternary"
      fullWidth={false}
      href={null}
      id="edit-card"
      onClick={[Function]}
      submit={false}
      variant="button"
    >
      Add new card
    </Button>
  </BaseStyle>
</section>
`;
